### ISSUE TITLE
label_image deadlocks

### ISSUE THREAD

**Issue by dvyukov** [tensorflow/tensorflow#929]
I am on commit 14cd77baeb0ee6f4b40816afc1f8c9c5d186bb8e.

I've added the following change that randomizes order of execution of tasks in thread pool:
```
diff --git a/tensorflow/core/lib/core/threadpool.cc b/tensorflow/core/lib/core/threadpool.cc
index 50aec3e..03c4485 100644
--- a/tensorflow/core/lib/core/threadpool.cc
+++ b/tensorflow/core/lib/core/threadpool.cc
@@ -20,6 +20,8 @@ limitations under the License.
 #include "tensorflow/core/platform/tracing.h"
 #include "tensorflow/core/platform/types.h"
 
+#include <unistd.h>
+
 namespace tensorflow {
 namespace thread {
 
@@ -34,6 +36,7 @@ ThreadPool::ThreadPool(Env* env, const string& name, int num_threads)
 ThreadPool::ThreadPool(Env* env, const ThreadOptions& thread_options,
                        const string& name, int num_threads)
     : name_(name) {
+  rand_ = getpid();
   CHECK_GE(num_threads, 1);
   string name_prefix = "tf_" + name_;
   for (int i = 0; i < num_threads; i++) {
@@ -102,6 +105,7 @@ void ThreadPool::WorkerLoop() {
         w.cv.wait(l);
       }
     }
+    std::swap(pending_[0], pending_[rand_r(&rand_) % pending_.size()]);
     // Pick up pending work
     Item item = pending_.front();
     pending_.pop_front();
diff --git a/tensorflow/core/lib/core/threadpool.h b/tensorflow/core/lib/core/threadpool.h
index ef37dcf..ae1eef3 100644
--- a/tensorflow/core/lib/core/threadpool.h
+++ b/tensorflow/core/lib/core/threadpool.h
@@ -66,6 +66,7 @@ class ThreadPool {
   std::vector<Thread*> threads_;  // All threads
   std::vector<Waiter*> waiters_;  // Stack of waiting threads.
   std::deque<Item> pending_;      // Queue of pending work
+  unsigned rand_;
 
   TF_DISALLOW_COPY_AND_ASSIGN(ThreadPool);
 };
```

Then run label_image as:
```
while echo OK; do bazel-bin/tensorflow/examples/label_image/label_image; done
```
after few iterations it deadlocks. All threads are blocked on condition variables, but tasks they are waiting for can't run because all thread pool threads are busy:

```
(gdb) info threads
  Id   Target Id         Frame 
  97   Thread 0x7f0d04038700 (LWP 74836) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  96   Thread 0x7f0d03837700 (LWP 74837) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  95   Thread 0x7f0d03036700 (LWP 74838) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  94   Thread 0x7f0d02835700 (LWP 74839) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  93   Thread 0x7f0d02034700 (LWP 74840) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  92   Thread 0x7f0d01833700 (LWP 74841) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  91   Thread 0x7f0d01032700 (LWP 74842) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  90   Thread 0x7f0d00831700 (LWP 74843) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  89   Thread 0x7f0cf3fff700 (LWP 74844) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  88   Thread 0x7f0cf37fe700 (LWP 74845) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  87   Thread 0x7f0cf2ffd700 (LWP 74846) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  86   Thread 0x7f0cf27fc700 (LWP 74847) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  85   Thread 0x7f0cf1ffb700 (LWP 74848) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  84   Thread 0x7f0cf17fa700 (LWP 74849) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  83   Thread 0x7f0cf0ff9700 (LWP 74850) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  82   Thread 0x7f0ce7fff700 (LWP 74851) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  81   Thread 0x7f0ce77fe700 (LWP 74852) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  80   Thread 0x7f0ce6ffd700 (LWP 74853) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  79   Thread 0x7f0ce67fc700 (LWP 74854) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  78   Thread 0x7f0ce5ffb700 (LWP 74855) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  77   Thread 0x7f0ce57fa700 (LWP 74856) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  76   Thread 0x7f0ce4ff9700 (LWP 74857) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  75   Thread 0x7f0cdffff700 (LWP 74858) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  74   Thread 0x7f0cdf7fe700 (LWP 74859) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  73   Thread 0x7f0cdeffd700 (LWP 74860) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  72   Thread 0x7f0cde7fc700 (LWP 74861) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  71   Thread 0x7f0cddffb700 (LWP 74862) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  70   Thread 0x7f0cdd7fa700 (LWP 74863) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  69   Thread 0x7f0cdcff9700 (LWP 74864) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  68   Thread 0x7f0cdc7f8700 (LWP 74865) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  67   Thread 0x7f0cdbff7700 (LWP 74866) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  66   Thread 0x7f0cdb7f6700 (LWP 74867) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  65   Thread 0x7f0cdaff5700 (LWP 74868) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  64   Thread 0x7f0cda7f4700 (LWP 74869) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  63   Thread 0x7f0cd9ff3700 (LWP 74870) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  62   Thread 0x7f0cd97f2700 (LWP 74871) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  61   Thread 0x7f0cd8ff1700 (LWP 74872) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  60   Thread 0x7f0cd3fff700 (LWP 74873) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  59   Thread 0x7f0cd37fe700 (LWP 74874) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  58   Thread 0x7f0cd2ffd700 (LWP 74875) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  57   Thread 0x7f0cd27fc700 (LWP 74876) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  56   Thread 0x7f0cd1ffb700 (LWP 74877) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  55   Thread 0x7f0cd17fa700 (LWP 74878) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  54   Thread 0x7f0cd0ff9700 (LWP 74879) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  53   Thread 0x7f0cd07f8700 (LWP 74880) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  52   Thread 0x7f0ccfff7700 (LWP 74881) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  51   Thread 0x7f0ccf7f6700 (LWP 74882) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  50   Thread 0x7f0cceff5700 (LWP 74883) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  49   Thread 0x7f0cce7f4700 (LWP 74884) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  48   Thread 0x7f0ccdff3700 (LWP 74885) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  47   Thread 0x7f0ccd7f2700 (LWP 74886) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  46   Thread 0x7f0cccff1700 (LWP 74887) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  45   Thread 0x7f0cc7fff700 (LWP 74888) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  44   Thread 0x7f0cc77fe700 (LWP 74889) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  43   Thread 0x7f0cc6ffd700 (LWP 74890) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  42   Thread 0x7f0cc67fc700 (LWP 74891) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  41   Thread 0x7f0cc5ffb700 (LWP 74892) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  40   Thread 0x7f0cc57fa700 (LWP 74893) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  39   Thread 0x7f0cc4ff9700 (LWP 74894) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  38   Thread 0x7f0cb7fff700 (LWP 74895) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  37   Thread 0x7f0cb77fe700 (LWP 74896) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  36   Thread 0x7f0cb6ffd700 (LWP 74897) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
---Type <return> to continue, or q <return> to quit---
  35   Thread 0x7f0cb67fc700 (LWP 74898) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  34   Thread 0x7f0cb5ffb700 (LWP 74899) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  33   Thread 0x7f0cb57fa700 (LWP 74900) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  32   Thread 0x7f0cb4ff9700 (LWP 74901) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  31   Thread 0x7f0caffff700 (LWP 74902) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  30   Thread 0x7f0caf7fe700 (LWP 74903) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  29   Thread 0x7f0caeffd700 (LWP 74904) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  28   Thread 0x7f0cae7fc700 (LWP 74905) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  27   Thread 0x7f0cadffb700 (LWP 74906) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  26   Thread 0x7f0cad7fa700 (LWP 74907) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  25   Thread 0x7f0cacff9700 (LWP 74908) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  24   Thread 0x7f0ca7fff700 (LWP 74909) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  23   Thread 0x7f0ca77fe700 (LWP 74910) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  22   Thread 0x7f0ca6ffd700 (LWP 74911) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  21   Thread 0x7f0ca67fc700 (LWP 74912) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  20   Thread 0x7f0ca5ffb700 (LWP 74913) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  19   Thread 0x7f0ca57fa700 (LWP 74914) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  18   Thread 0x7f0ca4ff9700 (LWP 74915) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  17   Thread 0x7f0ca47f8700 (LWP 74916) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  16   Thread 0x7f0ca3ff7700 (LWP 74917) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  15   Thread 0x7f0ca37f6700 (LWP 74918) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  14   Thread 0x7f0ca2ff5700 (LWP 74919) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  13   Thread 0x7f0ca27f4700 (LWP 74920) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  12   Thread 0x7f0ca1ff3700 (LWP 74921) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  11   Thread 0x7f0ca17f2700 (LWP 74922) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  10   Thread 0x7f0ca0ff1700 (LWP 74923) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  9    Thread 0x7f0c9bfff700 (LWP 74924) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  8    Thread 0x7f0c9b7fe700 (LWP 74925) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  7    Thread 0x7f0c9affd700 (LWP 74926) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  6    Thread 0x7f0c9a7fc700 (LWP 74927) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  5    Thread 0x7f0c99ffb700 (LWP 74928) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  4    Thread 0x7f0c997fa700 (LWP 74929) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  3    Thread 0x7f0c98ff9700 (LWP 74930) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  2    Thread 0x7f0c987f8700 (LWP 74931) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
* 1    Thread 0x7f0d0ad4c780 (LWP 74835) "label_image" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
(gdb) thread 53
[Switching to thread 53 (Thread 0x7f0cd07f8700 (LWP 74880))]
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
185	in ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S
(gdb) bt
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
#1  0x00007f0d09fb94bc in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>)
    at /build/buildd/gcc-4.8-4.8.4/build/x86_64-linux-gnu/libstdc++-v3/include/x86_64-linux-gnu/bits/gthr-default.h:864
#2  std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../src/libstdc++-v3/src/c++11/condition_variable.cc:52
#3  0x00000000004464fc in Eigen::Notification::WaitForNotification() ()
#4  0x000000000064f49a in void Eigen::TensorEvaluator<Eigen::TensorContractionOp<Eigen::array<Eigen::IndexPair<long>, 1ul> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16> const> const> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16> const> const> const, Eigen::ThreadPoolDevice>::packRhsAndKernel<Eigen::internal::packRhsAndKernelArg<float, float, Eigen::internal::TensorContractionInputMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16> const> const> const, Eigen::ThreadPoolDevice>, Eigen::array<long, 1ul>, Eigen::array<long, 1ul>, 4, true, false, 0>, Eigen::internal::blas_data_mapper<float, long, 0, 0>, long>, Eigen::internal::gemm_pack_rhs<float, long, Eigen::internal::TensorContractionSubMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16> const> const> const, Eigen::ThreadPoolDevice>, Eigen::array<long, 1ul>, Eigen::array<long, 1ul>, 4, true, false, 0>, 4, 0, false, false>, Eigen::internal::gebp_kernel<float, float, long, Eigen::internal::blas_data_mapper<float, long, 0, 0>, 8, 4, false, false> >(Eigen::internal::packRhsAndKernelArg<float, float, Eigen::internal::TensorContractionInputMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16> const> const> const, Eigen::ThreadPoolDevice>, Eigen::array<long, 1ul>, Eigen::array<long, 1ul>, 4, true, false, 0>, Eigen::internal::blas_data_mapper<float, long, 0, 0>, long>) ()
#5  0x0000000000634b1f in std::_Function_handler<void (), std::_Bind<void (*(Eigen::internal::packRhsAndKernelArg<float, float, Eigen::internal::TensorContractionInputMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16> const> const> const, Eigen::ThreadPoolDevice>, Eigen::array<long, 1ul>, Eigen::array<long, 1ul>, 4, true, false, 0>, Eigen::internal::blas_data_mapper<float, long, 0, 0>, long>))(Eigen::internal::packRhsAndKernelArg<float, float, Eigen::internal::TensorContractionInputMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16> const> const> const, Eigen::ThreadPoolDevice>, Eigen::array<long, 1ul>, Eigen::array<long, 1ul>, 4, true, false, 0>, Eigen::internal::blas_data_mapper<float, long, 0, 0>, long>)> >::_M_invoke(std::_Any_data const&) ()
#6  0x0000000000e44cef in std::_Function_handler<void (), tensorflow::thread::ThreadPoolDefaultImpl::ThreadPoolDefaultImpl(tensorflow::Env*, tensorflow::ThreadOptions const&, std::string const&, int)::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
#7  0x00007f0d09fbca40 in std::(anonymous namespace)::execute_native_thread_routine (__p=<optimized out>) at ../../../../../src/libstdc++-v3/src/c++11/thread.cc:84
#8  0x00007f0d0a217184 in start_thread (arg=0x7f0cd07f8700) at pthread_create.c:312
#9  0x00007f0d09a2a34d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

This come up during testing of a faster thread pool implementation that has distributed queues and so does not preserve FIFO order. But I think it can come up with current pool as well (maybe after some unrelated code changes, or maybe just due to unlucky scheduling order).

Computational tasks running on a fixed-size thread pool must not ever block. Instead they should schedule continuations. Besides deadlocks blocking leads to serious CPU underutilization. I.e. on my machine label_image utilizes only about half of cores, because half of threads in pool are blocked waiting for other tasks.

**Comment by dvyukov** [comment]
@jeremybarnes I wonder didn't you this with your pool?

**Comment by jeremybarnes** [comment]
(I've been submerged for the last couple of weeks, and I didn't see the notification).

I didn't see any deadlocks, but I can see that they are definitely possible on any thread pool that doesn't know to schedule a new thread if a worker thread blocks.  We need to be very careful in any bounded thread scenario, as it's not possible to guarantee progress in the general case on a finite number of threads.

In general the pattern I use to work around these kind of deadlocks with a fixed number of threads is to poll instead of block, and perform work from the threadpool after un unsuccessful poll to guarantee forward progress.  That would require some deeper changes to Eigen so that it knew it was in a worker thread and could adjust its behaviour accordingly.

**Comment by dvyukov** [comment]
Thanks, Jeremy. To clarify, you run label_image with your pool and did not see deadlocks; or you did not run label_image and did not see deadlocks?

Pulling for work while waiting can help to some degree, but it still has problems when blocked tasks stack up on a single thread:
1. Stack can grow without limit and cause stack overflow, and you can't limit recursion at runtime because then you get back to deadlocks.
2. This limit available parallelism: a task can be already ready to run but is sitting on a thread stack, and the thread is busy executing another task. It is not possible to execute the readied task until all tasks on the same thread finish.
3. And more importantly, it is subject to deadlocks as well. Consider that a task is waiting for completion of a task that is blocked on the very same stack below it. Oops.

Some parallel computation libraries avoid the deadlocks by imposing a DAG requirement of work structure where all dependent tasks must be spawned by the task itself. This again requires fixing structure of all algorithms rather than just plugging polling into wait objects. And it makes (2) worse.

Some other systems solve this by playing tricks with virtual memory:
http://groups.csail.mit.edu/sct/wiki/index.php?title=The_Cilk-M_Project
http://supertech.csail.mit.edu/papers/stacks.pdf
But I think this is out of question here.

I would suggest to explicitly schedule continuations instead of blocking on worker thread stack. Continuations suck code-wise. But that's what we have in C++. Hopefully it's not that bad with closures.

**Comment by jeremybarnes** [comment]
I ran label_image and did not see any deadlocks, although I wasn't looking for them.

I don't think that continuations would be an issue code-wise; it's more enabling the synchronization methods to accept them that would take work.

**Comment by jeremybarnes** [comment]
Also, the reason for having work-stealing across thread's work queues was to deal with your point 2 above.  I agree it's a big deal.

**Comment by alquraishi** [comment]
An issue I've been having with multiple file and batch queues may be related. See this [question](http://stackoverflow.com/questions/35414009/multiple-queues-causing-tf-to-lock-up/35420394#35420394) on stack overflow.

**Comment by girving** [comment]
@dvyukov, @rmlarsen: Did the recent thread pool changes resolve this?

**Comment by dvyukov** [comment]
Yes, it's fixed with the contraction changes.
